File: packages/core/semantic/authoring.ts
import { z } from "zod";
import { queryOptions } from "@tanstack/react-query";
import { semanticRequest } from "./api";
import { semanticKeys } from "./queries";
import { coordinatorResumeSchema } from "./coordinator-resume";

const questionSchema = z
  .object({
    key: z.string(),
    topic: z.string(),
    prompt: z.string(),
    why: z.string(),
    status: z.enum(["unanswered", "answered", "skipped"]),
    answer: z.string().optional(),
    answer_kind: z.string().optional(),
    answered_by: z.string().optional(),
    decision_id: z.string().optional(),
  })
  .transform((v) => ({
    key: v.key,
    topic: v.topic,
    prompt: v.prompt,
    why: v.why,
    status: v.status,
    answer: v.answer,
    answerKind: v.answer_kind,
    answeredBy: v.answered_by,
    decisionId: v.decision_id,
  }));
const cardSchema = z
  .object({
    key: z.string(),
    kind: z.enum([
      "entity",
      "attribute",
      "relationship",
      "action",
      "policy",
      "data_binding",
      "action_binding",
    ]),
    label: z.string(),
    description: z.string(),
    classification: z.string(),
    source_refs: z.array(z.string()).catch([]),
    status: z.enum([
      "proposed",
      "confirmed",
      "corrected",
      "unknown",
      "rejected",
    ]),
    rationale: z.string().optional(),
    decided_by: z.string().optional(),
    decision_id: z.string().optional(),
  })
  .transform((v) => ({
    key: v.key,
    kind: v.kind,
    label: v.label,
    description: v.description,
    classification: v.classification,
    sourceRefs: v.source_refs,
    status: v.status,
    rationale: v.rationale,
    decidedBy: v.decided_by,
    decisionId: v.decision_id,
  }));
export const authoringSchema = z
  .object({
    construction_id: z.string(),
    revision: z.number().int().nonnegative(),
    interview: z.object({
      scope: z.record(z.string(), z.unknown()).optional(),
      status: z.enum(["collecting", "ready"]),
      round: z.number(),
      questions: z.array(questionSchema),
    }),
    cards: z.array(cardSchema),
    coordinator_resume: coordinatorResumeSchema.nullish(),
  })
  .transform((v) => ({
    constructionId: v.construction_id,
    coordinatorResume: v.coordinator_resume ?? undefined,
    revision: v.revision,
    interview: v.interview,
    cards: v.cards,
  }));
export type OntologyAuthoring = z.infer<typeof authoringSchema>;
export type AuthoringCard = OntologyAuthoring["cards"][number];
export type AuthoringResponse = {
  expectedRevision: number;
  answers?: {
    questionKey: string;
    answer: string;
    answerKind?: "answer" | "unknown";
  }[];
  cardDecisions?: {
    cardKey: string;
    decision: "confirm" | "correct" | "unknown" | "reject";
    rationale: string;
    correction?: {
      kind?: AuthoringCard["kind"];
      label?: string;
      description?: string;
    };
  }[];
};
export const authoringApi = {
  forIssue: async (issueId: string) => {
    try {
      return await semanticRequest(
        `/constructions/for-issue/${encodeURIComponent(issueId)}`,
        z
          .object({ construction_id: z.string() })
          .transform((v) => ({ constructionId: v.construction_id })),
      );
    } catch (error) {
      if (error instanceof Error && "status" in error && error.status === 404)
        return null;
      throw error;
    }
  },
  get: (constructionId: string) =>
    semanticRequest(
      `/constructions/${encodeURIComponent(constructionId)}/authoring`,
      authoringSchema,
    ),
  respond: (constructionId: string, input: AuthoringResponse) =>
    semanticRequest(
      `/constructions/${encodeURIComponent(constructionId)}/authoring/respond`,
      authoringSchema,
      {
        method: "POST",
        body: {
          expected_revision: input.expectedRevision,
          answers: input.answers?.map((v) => ({
            question_key: v.questionKey,
            answer: v.answer,
            answer_kind: v.answerKind,
          })),
          card_decisions: input.cardDecisions?.map((v) => ({
            card_key: v.cardKey,
            decision: v.decision,
            rationale: v.rationale,
            correction: v.correction,
          })),
        },
      },
    ),
};
export const issueConstructionOptions = (wsId: string, issueId: string) =>
  queryOptions({
    queryKey: [...semanticKeys.all(wsId), "issue-construction", issueId],
    queryFn: () => authoringApi.forIssue(issueId),
    enabled: !!issueId,
    retry: false,
    refetchInterval: 10000,
  });
export const authoringOptions = (wsId: string, constructionId: string) =>
  queryOptions({
    queryKey: [...semanticKeys.all(wsId), "authoring", constructionId],
    queryFn: () => authoringApi.get(constructionId),
    enabled: !!constructionId,
    refetchInterval: 5000,
  });

File: server/internal/ontologizer/authoring.go
package ontologizer

import (
	"errors"
	"strings"
)

type InterviewState struct {
	Scope     map[string]any      `json:"scope,omitempty"`
	Status    string              `json:"status"`
	Round     int                 `json:"round"`
	Questions []InterviewQuestion `json:"questions"`
}

type InterviewQuestion struct {
	Key        string `json:"key"`
	Topic      string `json:"topic"`
	Prompt     string `json:"prompt"`
	Why        string `json:"why"`
	Status     string `json:"status"`
	Answer     string `json:"answer,omitempty"`
	AnswerKind string `json:"answer_kind,omitempty"`
	AnsweredBy string `json:"answered_by,omitempty"`
	DecisionID string `json:"decision_id,omitempty"`
}

type CandidateCard struct {
	Key            string   `json:"key"`
	Kind           string   `json:"kind"`
	Label          string   `json:"label"`
	Description    string   `json:"description"`
	Classification string   `json:"classification"`
	SourceRefs     []string `json:"source_refs,omitempty"`
	Status         string   `json:"status"`
	Rationale      string   `json:"rationale,omitempty"`
	DecidedBy      string   `json:"decided_by,omitempty"`
	DecisionID     string   `json:"decision_id,omitempty"`
}

func ValidateInterviewProposal(state InterviewState) error {
	if state.Round < 0 || len(state.Questions) > 5 {
		return errors.New("an interview proposal may contain at most five unanswered questions")
	}
	switch state.Status {
	case "collecting", "ready":
	default:
		return errors.New("interview status must be collecting or ready")
	}
	seen := map[string]bool{}
	for _, question := range state.Questions {
		if strings.TrimSpace(question.Key) == "" || strings.TrimSpace(question.Prompt) == "" || strings.TrimSpace(question.Why) == "" {
			return errors.New("interview questions require key, prompt and why")
		}
		if seen[question.Key] {
			return errors.New("interview question keys must be unique")
		}
		seen[question.Key] = true
		switch question.Topic {
		case "goal", "boundary", "decisions", "objects", "process", "exceptions", "terminology", "granularity":
		default:
			return errors.New("interview question has an invalid topic")
		}
		if question.Answer != "" || question.AnswerKind != "" || question.AnsweredBy != "" || question.DecisionID != "" {
			return errors.New("agents cannot write human interview answers")
		}
		if question.Status != "" && question.Status != "unanswered" {
			return errors.New("agents may only propose unanswered questions")
		}
	}
	return nil
}

func ValidateCandidateCardProposal(cards []CandidateCard) error {
	seen := map[string]bool{}
	for _, card := range cards {
		if strings.TrimSpace(card.Key) == "" || strings.TrimSpace(card.Label) == "" || strings.TrimSpace(card.Description) == "" {
			return errors.New("candidate cards require key, label and description")
		}
		if seen[card.Key] {
			return errors.New("candidate card keys must be unique")
		}
		seen[card.Key] = true
		switch card.Kind {
		case "entity", "attribute", "relationship", "action", "policy", "data_binding", "action_binding":
		default:
			return errors.New("candidate card kind must be an ontology element or binding")
		}
		switch card.Classification {
		case "fact":
			if len(card.SourceRefs) == 0 {
				return errors.New("fact cards require source references")
			}
		case "recommendation", "unknown":
		default:
			return errors.New("agents may classify cards only as fact, recommendation or unknown")
		}
		if card.Status != "" && card.Status != "proposed" || card.Rationale != "" || card.DecidedBy != "" || card.DecisionID != "" {
			return errors.New("agents cannot decide candidate cards")
		}
	}
	return nil
}
